Source/Controllers/Other/SQLitePinnedTableManager.swift

```diff
@@ -0,0 +1,207 @@
+
+                while rs.next() {
+                    let hostName = rs.string(forColumn: "hostname")!
+                    let databaseName = rs.string(forColumn: "databaseName")!
+                    let pinnedTableName = rs.string(forColumn: "pinnedTableName")!
+                    addToPinnedTablesDatabaseDictionary(hostName: hostName, databaseName: databaseName, tableToPin: pinnedTableName)
+                }
+                rs.close()
+            } catch {
+                logDBError(error)
+            }
+        }
+        queue.close()
+    }
+
+    @objc func getPinnedTables(hostName: String, databaseName: String) -> [String] {
+        return pinnedTablesDatabaseDictionary[hostName]?[databaseName] ?? []
+    }
+
+    @objc func pinTable(hostName: String, databaseName: String, tableToPin: String) {
+        
+        if (pinnedTablesDatabaseDictionary[hostName]?[databaseName] != nil && pinnedTablesDatabaseDictionary[hostName]![databaseName]!.contains(tableToPin)) {
+            return // tableToPin already pinned
+        }
+        addToPinnedTablesDatabaseDictionary(hostName: hostName, databaseName: databaseName, tableToPin: tableToPin)
+        queue.inDatabase { db in
+            db.traceExecution = traceExecution
+            do {
+                try db.executeUpdate("INSERT INTO PinnedTables (hostName, databaseName, pinnedTableName) VALUES (?, ?, ?)",
+                        values: [hostName, databaseName, tableToPin])
+            } catch {
+                logDBError(error)
+            }
+        }
+        queue.close()
+    }
+    
+
+    @objc func unpinTable(hostName: String, databaseName: String, tableToUnpin: String) {
+        if pinnedTablesDatabaseDictionary[hostName]?[databaseName] == nil {
+            return
+        }
+        if !pinnedTablesDatabaseDictionary[hostName]![databaseName]!.contains(tableToUnpin) {
+            return
+        }
+        let index = pinnedTablesDatabaseDictionary[hostName]![databaseName]!.firstIndex(of: tableToUnpin)!
+        pinnedTablesDatabaseDictionary[hostName]![databaseName]!.remove(at: index)
+
+        queue.inDatabase { db in
+            db.traceExecution = traceExecution
+            do {
+                try db.executeUpdate("DELETE FROM PinnedTables where hostName=? and databaseName=? and pinnedTableName=?",
+                        values: [hostName, databaseName, tableToUnpin])
+            } catch {
+                logDBError(error)
+            }
+        }
+        queue.close()
+    }
+    
+    
+    private func addToPinnedTablesDatabaseDictionary(hostName: String, databaseName: String, tableToPin: String) {
+        if pinnedTablesDatabaseDictionary[hostName] == nil {
+            pinnedTablesDatabaseDictionary[hostName] = [:];
+        }
+        if pinnedTablesDatabaseDictionary[hostName]?[databaseName] == nil {
```

Comment:
^^ can we work with local var instead of accessing all this all the time?

Reply:
By working with local variable, you mean something like this?
`if pinnedTablesDatabaseDictionary[hostName] == nil {
            pinnedTablesDatabaseDictionary[hostName] = [:];
        }
        var pinnedTablesForHost = pinnedTablesDatabaseDictionary[hostName];
        if pinnedTablesForHost?[databaseName] == nil {
            pinnedTablesForHost?[databaseName] = []
        }
        var pinnedTablesForHostAndDatabase = pinnedTablesForHost?[databaseName];
        pinnedTablesForHostAndDatabase?.append(tableToPin)`